File: services/ml-service/src/domain/training/runner.rs
use serde_json::{Value, json};

use crate::{
    domain::interop,
    models::{run::MetricValue, training_job::TrainingTrial},
};

use super::hyperparameter;

#[derive(Debug, Clone)]
pub struct TrialOutcome {
    pub trial: TrainingTrial,
    pub metrics: Vec<MetricValue>,
    pub schema: Value,
}

#[derive(Debug, Clone)]
struct TrainingDataset {
    feature_names: Vec<String>,
    feature_means: Vec<f64>,
    feature_scales: Vec<f64>,
    rows: Vec<Vec<f64>>,
    labels: Vec<f64>,
    label_field: String,
    positive_label: String,
    negative_label: String,
}

pub fn has_inline_training_data(training_config: &Value) -> bool {
    training_config
        .get("records")
        .and_then(Value::as_array)
        .map(|records| !records.is_empty())
        .unwrap_or(false)
}

pub fn train_trial(
    training_config: &Value,
    hyperparameters: &Value,
    objective_metric_name: &str,
    trial_index: usize,
) -> Result<TrialOutcome, String> {
    let dataset = parse_dataset(training_config)?;
    let learning_rate = hyperparameter::value_as_f64(hyperparameters.get("learning_rate"), 0.08);
    let epochs = hyperparameter::value_as_u64(hyperparameters.get("epochs"), 350) as usize;
    let l2 = hyperparameter::value_as_f64(hyperparameters.get("l2"), 0.0);

    let (weights, bias) =
        fit_logistic_regression(&dataset.rows, &dataset.labels, learning_rate, epochs, l2);
    let metrics = evaluate_metrics(&dataset, &weights, bias);
    let objective_metric =
        select_metric(&metrics, objective_metric_name).unwrap_or_else(|| metrics[0].clone());

    let schema = json!({
        "signature": "tabular-binary",
        "engine": interop::effective_framework(training_config),
        "model_adapter": crate::handlers::to_json(&interop::infer_model_adapter(
            Some(training_config),
            None,
        )),
        "model_state": {
            "feature_names": dataset.feature_names,
            "feature_means": dataset.feature_means,
            "feature_scales": dataset.feature_scales,
            "weights": weights,
            "bias": bias,
            "threshold": 0.5,
            "label_field": dataset.label_field,
            "positive_label": dataset.positive_label,
            "negative_label": dataset.negative_label,
        },
        "training_summary": {
            "row_count": dataset.rows.len(),
            "feature_count": dataset.rows.first().map(|row| row.len()).unwrap_or(0),
            "objective_metric": objective_metric.name,
            "objective_value": objective_metric.value,
            "framework": interop::effective_framework(training_config),
        }
    });

    Ok(TrialOutcome {
        trial: TrainingTrial {
            id: format!("trial-{}", trial_index + 1),
            status: "completed".to_string(),
            hyperparameters: hyperparameters.clone(),
            objective_metric,
        },
        metrics,
        schema,
    })
}

fn parse_dataset(training_config: &Value) -> Result<TrainingDataset, String> {
    let records = training_config
        .get("records")
        .and_then(Value::as_array)
        .ok_or_else(|| "training_config.records must be a non-empty array".to_string())?;
    if records.is_empty() {
        return Err("training_config.records must contain at least one row".to_string());
    }

    let label_field = training_config
        .get("label_field")
        .and_then(Value::as_str)
        .unwrap_or("label")
        .to_string();
    let positive_label = training_config
        .get("positive_label")
        .and_then(Value::as_str)
        .unwrap_or("positive")
        .to_string();
    let negative_label = training_config
        .get("negative_label")
        .and_then(Value::as_str)
        .unwrap_or("negative")
        .to_string();

    let feature_names = training_config
        .get("features")
        .and_then(Value::as_array)
        .map(|features| {
            features
                .iter()
                .filter_map(Value::as_str)
                .map(ToOwned::to_owned)
                .collect::<Vec<_>>()
        })
        .filter(|features| !features.is_empty())
        .unwrap_or_else(|| derive_feature_names(records, &label_field));

    if feature_names.is_empty() {
        return Err("training_config.features resolved to an empty set".to_string());
    }

    let mut raw_rows = Vec::with_capacity(records.len());
    let mut labels = Vec::with_capacity(records.len());

    for record in records {
        let object = record
            .as_object()
            .ok_or_else(|| "each training record must be a JSON object".to_string())?;
        let label_value = object
            .get(&label_field)
            .ok_or_else(|| format!("missing label field '{label_field}'"))?;
        labels.push(binary_label(label_value, &positive_label));
        raw_rows.push(
            feature_names
                .iter()
                .map(|feature| scalar_feature(object.get(feature)))
                .collect::<Vec<_>>(),
        );
    }

    let (feature_means, feature_scales, rows) = standardize_rows(&raw_rows);

    Ok(TrainingDataset {
        feature_names,
        feature_means,
        feature_scales,
        rows,
        labels,
        label_field,
        positive_label,
        negative_label,
    })
}

fn derive_feature_names(records: &[Value], label_field: &str) -> Vec<String> {
    let mut names = records
        .iter()
        .filter_map(Value::as_object)
        .flat_map(|object| object.keys().cloned())
        .filter(|name| name != label_field)
        .collect::<Vec<_>>();
    names.sort();
    names.dedup();
    names
}

fn scalar_feature(value: Option<&Value>) -> f64 {
    match value {
        Some(Value::Number(number)) => number.as_f64().unwrap_or(0.0),
        Some(Value::Bool(flag)) => {
            if *flag {
                1.0
            } else {
                0.0
            }
        }
        Some(Value::String(text)) => text.parse::<f64>().unwrap_or_else(|_| {
            let hash = text
                .bytes()
                .fold(0u64, |acc, byte| acc.wrapping_add(byte as u64));
            (hash % 1000) as f64 / 1000.0
        }),
        _ => 0.0,
    }
}

fn binary_label(value: &Value, positive_label: &str) -> f64 {
    match value {
        Value::Bool(flag) => {
            if *flag {
                1.0
            } else {
                0.0
            }
        }
        Value::Number(number) => {
            if number.as_f64().unwrap_or(0.0) >= 0.5 {
                1.0
            } else {
                0.0
            }
        }
        Value::String(text) => {
            if text == positive_label || text.eq_ignore_ascii_case("true") || text == "1" {
                1.0
            } else {
                0.0
            }
        }
        _ => 0.0,
    }
}

fn standardize_rows(rows: &[Vec<f64>]) -> (Vec<f64>, Vec<f64>, Vec<Vec<f64>>) {
    let feature_count = rows.first().map(|row| row.len()).unwrap_or(0);
    let mut means = vec![0.0; feature_count];
    let mut scales = vec![1.0; feature_count];

    if rows.is_empty() {
        return (means, scales, Vec::new());
    }

    for row in rows {
        for (index, value) in row.iter().enumerate() {
            means[index] += *value;
        }
    }
    for mean in &mut means {
        *mean /= rows.len() as f64;
    }

    for row in rows {
        for (index, value) in row.iter().enumerate() {
            let delta = *value - means[index];
            scales[index] += delta * delta;
        }
    }
    for scale in &mut scales {
        *scale = (*scale / rows.len() as f64).sqrt();
        if *scale == 0.0 {
            *scale = 1.0;
        }
    }

    let standardized = rows
        .iter()
        .map(|row| {
            row.iter()
                .enumerate()
                .map(|(index, value)| (*value - means[index]) / scales[index])
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();

    (means, scales, standardized)
}

fn fit_logistic_regression(
    rows: &[Vec<f64>],
    labels: &[f64],
    learning_rate: f64,
    epochs: usize,
    l2: f64,
) -> (Vec<f64>, f64) {
    let feature_count = rows.first().map(|row| row.len()).unwrap_or(0);
    let mut weights = vec![0.0; feature_count];
    let mut bias = 0.0;

    if rows.is_empty() {
        return (weights, bias);
    }

    for _ in 0..epochs {
        let mut gradient = vec![0.0; feature_count];
        let mut bias_gradient = 0.0;

        for (row, label) in rows.iter().zip(labels.iter()) {
            let prediction = sigmoid(dot(&weights, row) + bias);
            let error = prediction - *label;

            for (index, value) in row.iter().enumerate() {
                gradient[index] += error * value;
            }
            bias_gradient += error;
        }

        let row_count = rows.len() as f64;
        for (index, weight) in weights.iter_mut().enumerate() {
            let regularized = gradient[index] / row_count + l2 * *weight;
            *weight -= learning_rate * regularized;
        }
        bias -= learning_rate * (bias_gradient / row_count);
    }

    (weights, bias)
}

fn evaluate_metrics(dataset: &TrainingDataset, weights: &[f64], bias: f64) -> Vec<MetricValue> {
    let mut true_positive = 0.0_f64;
    let mut true_negative = 0.0_f64;
    let mut false_positive = 0.0_f64;
    let mut false_negative = 0.0_f64;
    let mut log_loss = 0.0_f64;

    for (row, label) in dataset.rows.iter().zip(dataset.labels.iter()) {
        let probability = sigmoid(dot(weights, row) + bias).clamp(1e-6, 1.0 - 1e-6);
        let predicted = if probability >= 0.5 { 1.0 } else { 0.0 };
        log_loss += -(*label * probability.ln() + (1.0 - *label) * (1.0 - probability).ln());

        match (predicted, *label) {
            (1.0, 1.0) => true_positive += 1.0,
            (0.0, 0.0) => true_negative += 1.0,
            (1.0, 0.0) => false_positive += 1.0,
            (0.0, 1.0) => false_negative += 1.0,
            _ => {}
        }
    }

    let total = (true_positive + true_negative + false_positive + false_negative).max(1.0_f64);
    let accuracy = round_metric((true_positive + true_negative) / total);
    let precision = round_metric(true_positive / (true_positive + false_positive).max(1.0));
    let recall = round_metric(true_positive / (true_positive + false_negative).max(1.0));
    let f1 = round_metric(if precision + recall == 0.0 {
        0.0
    } else {
        2.0 * precision * recall / (precision + recall)
    });

    vec![
        MetricValue {
            name: "accuracy".to_string(),
            value: accuracy,
        },
        MetricValue {
            name: "precision".to_string(),
            value: precision,
        },
        MetricValue {
            name: "recall".to_string(),
            value: recall,
        },
        MetricValue {
            name: "f1".to_string(),
            value: f1,
        },
        MetricValue {
            name: "log_loss".to_string(),
            value: round_metric(log_loss / total),
        },
    ]
}

fn select_metric(metrics: &[MetricValue], name: &str) -> Option<MetricValue> {
    metrics.iter().find(|metric| metric.name == name).cloned()
}

fn dot(left: &[f64], right: &[f64]) -> f64 {
    left.iter()
        .zip(right.iter())
        .map(|(l, r)| l * r)
        .sum::<f64>()
}

fn sigmoid(value: f64) -> f64 {
    1.0 / (1.0 + (-value).exp())
}

fn round_metric(value: f64) -> f64 {
    (value * 10_000.0).round() / 10_000.0
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{has_inline_training_data, train_trial};

    #[test]
    fn trains_real_logistic_trial_from_inline_records() {
        let config = json!({
            "engine": "tabular-logistic",
            "label_field": "label",
            "positive_label": "positive",
            "records": [
                { "label": "positive", "tickets_open": 9, "usage_delta": -0.8, "nps": 2 },
                { "label": "positive", "tickets_open": 7, "usage_delta": -0.6, "nps": 3 },
                { "label": "negative", "tickets_open": 1, "usage_delta": 0.2, "nps": 9 },
                { "label": "negative", "tickets_open": 2, "usage_delta": 0.4, "nps": 8 }
            ]
        });

        assert!(has_inline_training_data(&config));
        let outcome = train_trial(
            &config,
            &json!({ "learning_rate": 0.1, "epochs": 400, "l2": 0.0 }),
            "f1",
            0,
        )
        .unwrap();

        assert_eq!(outcome.trial.status, "completed");
        assert!(outcome.trial.objective_metric.value >= 0.8);
        assert!(outcome.schema.pointer("/model_state/weights").is_some());
    }
}

File: libs/ml-kernel-go/domain/training/runner.go
package training

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"

	"github.com/openfoundry/openfoundry-go/libs/ml-kernel-go/domain/interop"
	"github.com/openfoundry/openfoundry-go/libs/ml-kernel-go/models"
)

// TrialOutcome bundles a TrainingTrial with the per-trial metrics
// list + the derived model schema. Mirrors Rust TrialOutcome.
type TrialOutcome struct {
	Trial   models.TrainingTrial
	Metrics []models.MetricValue
	Schema  json.RawMessage
}

// trainingDataset is the parsed in-memory dataset the logistic
// regressor consumes. Private — exposed only via TrainTrial.
type trainingDataset struct {
	FeatureNames  []string
	FeatureMeans  []float64
	FeatureScales []float64
	Rows          [][]float64
	Labels        []float64
	LabelField    string
	PositiveLabel string
	NegativeLabel string
}

// HasInlineTrainingData mirrors fn has_inline_training_data — true
// when training_config.records is a non-empty array.
func HasInlineTrainingData(trainingConfig json.RawMessage) bool {
	if len(trainingConfig) == 0 {
		return false
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(trainingConfig, &obj); err != nil {
		return false
	}
	rawRecords, ok := obj["records"]
	if !ok {
		return false
	}
	var records []json.RawMessage
	if err := json.Unmarshal(rawRecords, &records); err != nil {
		return false
	}
	return len(records) > 0
}

// TrainTrial runs one logistic-regression training trial against the
// inline dataset in training_config. Mirrors fn train_trial verbatim.
func TrainTrial(trainingConfig, hyperparameters json.RawMessage, objectiveMetricName string, trialIndex int) (TrialOutcome, error) {
	dataset, err := parseDataset(trainingConfig)
	if err != nil {
		return TrialOutcome{}, err
	}

	hyper := map[string]any{}
	if len(hyperparameters) > 0 {
		_ = json.Unmarshal(hyperparameters, &hyper)
	}
	learningRate := ValueAsFloat64(hyper["learning_rate"], 0.08)
	epochs := int(ValueAsUint64(hyper["epochs"], 350))
	l2 := ValueAsFloat64(hyper["l2"], 0.0)

	weights, bias := fitLogisticRegression(dataset.Rows, dataset.Labels, learningRate, epochs, l2)
	metrics := evaluateMetrics(dataset, weights, bias)

	objectiveMetric := selectMetric(metrics, objectiveMetricName)
	if objectiveMetric == nil && len(metrics) > 0 {
		first := metrics[0]
		objectiveMetric = &first
	}
	if objectiveMetric == nil {
		objectiveMetric = &models.MetricValue{Name: objectiveMetricName, Value: 0}
	}

	framework := interop.EffectiveFramework(trainingConfig)
	adapter := interop.InferModelAdapter(trainingConfig, nil)

	rowCount := len(dataset.Rows)
	featureCount := 0
	if rowCount > 0 {
		featureCount = len(dataset.Rows[0])
	}

	schemaObj := map[string]any{
		"signature":     "tabular-binary",
		"engine":        framework,
		"model_adapter": adapter,
		"model_state": map[string]any{
			"feature_names":  dataset.FeatureNames,
			"feature_means":  dataset.FeatureMeans,
			"feature_scales": dataset.FeatureScales,
			"weights":        weights,
			"bias":           bias,
			"threshold":      0.5,
			"label_field":    dataset.LabelField,
			"positive_label": dataset.PositiveLabel,
			"negative_label": dataset.NegativeLabel,
		},
		"training_summary": map[string]any{
			"row_count":        rowCount,
			"feature_count":    featureCount,
			"objective_metric": objectiveMetric.Name,
			"objective_value":  objectiveMetric.Value,
			"framework":        framework,
		},
	}
	schemaJSON, _ := json.Marshal(schemaObj)

	hyperCopy := json.RawMessage(append([]byte(nil), hyperparameters...))
	if len(hyperCopy) == 0 {
		hyperCopy = json.RawMessage("{}")
	}

	return TrialOutcome{
		Trial: models.TrainingTrial{
			ID:              fmt.Sprintf("trial-%d", trialIndex+1),
			Status:          "completed",
			Hyperparameters: hyperCopy,
			ObjectiveMetric: *objectiveMetric,
		},
		Metrics: metrics,
		Schema:  schemaJSON,
	}, nil
}

func parseDataset(trainingConfig json.RawMessage) (*trainingDataset, error) {
	if len(trainingConfig) == 0 {
		return nil, errors.New("training_config.records must be a non-empty array")
	}
	var configObj map[string]json.RawMessage
	if err := json.Unmarshal(trainingConfig, &configObj); err != nil {
		return nil, errors.New("training_config.records must be a non-empty array")
	}
	rawRecords, ok := configObj["records"]
	if !ok {
		return nil, errors.New("training_config.records must be a non-empty array")
	}
	var records []map[string]any
	if err := json.Unmarshal(rawRecords, &records); err != nil {
		return nil, errors.New("each training record must be a JSON object")
	}
	if len(records) == 0 {
		return nil, errors.New("training_config.records must contain at least one row")
	}

	labelField := stringFromConfig(configObj, "label_field", "label")
	positiveLabel := stringFromConfig(configObj, "positive_label", "positive")
	negativeLabel := stringFromConfig(configObj, "negative_label", "negative")

	featureNames := stringArrayFromConfig(configObj, "features")
	if len(featureNames) == 0 {
		featureNames = deriveFeatureNames(records, labelField)
	}
	if len(featureNames) == 0 {
		return nil, errors.New("training_config.features resolved to an empty set")
	}

	rawRows := make([][]float64, 0, len(records))
	labels := make([]float64, 0, len(records))
	for _, record := range records {
		labelVal, has := record[labelField]
		if !has {
			return nil, fmt.Errorf("missing label field '%s'", labelField)
		}
		labels = append(labels, binaryLabel(labelVal, positiveLabel))
		row := make([]float64, 0, len(featureNames))
		for _, feature := range featureNames {
			row = append(row, scalarFeature(record[feature]))
		}
		rawRows = append(rawRows, row)
	}

	means, scales, rows := standardizeRows(rawRows)

	return &trainingDataset{
		FeatureNames:  featureNames,
		FeatureMeans:  means,
		FeatureScales: scales,
		Rows:          rows,
		Labels:        labels,
		LabelField:    labelField,
		PositiveLabel: positiveLabel,
		NegativeLabel: negativeLabel,
	}, nil
}

func stringFromConfig(obj map[string]json.RawMessage, key, fallback string) string {
	raw, ok := obj[key]
	if !ok || len(raw) == 0 {
		return fallback
	}
	var s string
	if err := json.Unmarshal(raw, &s); err != nil || s == "" {
		return fallback
	}
	return s
}

func stringArrayFromConfig(obj map[string]json.RawMessage, key string) []string {
	raw, ok := obj[key]
	if !ok || len(raw) == 0 {
		return nil
	}
	var arr []any
	if err := json.Unmarshal(raw, &arr); err != nil {
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, item := range arr {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func deriveFeatureNames(records []map[string]any, labelField string) []string {
	seen := map[string]struct{}{}
	names := []string{}
	for _, record := range records {
		for k := range record {
			if k == labelField {
				continue
			}
			if _, has := seen[k]; has {
				continue
			}
			seen[k] = struct{}{}
			names = append(names, k)
		}
	}
	sort.Strings(names)
	return names
}

func scalarFeature(v any) float64 {
	switch x := v.(type) {
	case float64:
		return x
	case float32:
		return float64(x)
	case int:
		return float64(x)
	case int32:
		return float64(x)
	case int64:
		return float64(x)
	case json.Number:
		if f, err := x.Float64(); err == nil {
			return f
		}
	case bool:
		if x {
			return 1.0
		}
		return 0.0
	case string:
		if f, err := strconv.ParseFloat(x, 64); err == nil {
			return f
		}
		var hash uint64
		for _, b := range []byte(x) {
			hash += uint64(b)
		}
		return float64(hash%1000) / 1000.0
	}
	return 0.0
}

func binaryLabel(v any, positiveLabel string) float64 {
	switch x := v.(type) {
	case bool:
		if x {
			return 1.0
		}
		return 0.0
	case float64:
		if x >= 0.5 {
			return 1.0
		}
		return 0.0
	case float32:
		if x >= 0.5 {
			return 1.0
		}
		return 0.0
	case int:
		if x >= 1 {
			return 1.0
		}
		return 0.0
	case int32:
		if x >= 1 {
			return 1.0
		}
		return 0.0
	case int64:
		if x >= 1 {
			return 1.0
		}
		return 0.0
	case json.Number:
		if f, err := x.Float64(); err == nil && f >= 0.5 {
			return 1.0
		}
		return 0.0
	case string:
		if x == positiveLabel || strings.EqualFold(x, "true") || x == "1" {
			return 1.0
		}
		return 0.0
	}
	return 0.0
}

func standardizeRows(rows [][]float64) (means, scales []float64, standardized [][]float64) {
	featureCount := 0
	if len(rows) > 0 {
		featureCount = len(rows[0])
	}
	means = make([]float64, featureCount)
	scales = make([]float64, featureCount)
	for i := range scales {
		scales[i] = 1.0
	}
	if len(rows) == 0 {
		return means, scales, [][]float64{}
	}

	for _, row := range rows {
		for i, v := range row {
			means[i] += v
		}
	}
	for i := range means {
		means[i] /= float64(len(rows))
	}

	// Reset scales to 0 before accumulating squared deltas (Rust
	// initialises scales=1 then *adds* delta² which is the same
	// numerical sequence as 0 + sum + ÷ N + sqrt because the +1
	// initial cancels with the same /N — actually no, the Rust code
	// uses += and starts at 1.0 so we must mirror it exactly).
	// Reset to 1.0 to track the +=.
	for i := range scales {
		scales[i] = 1.0
	}
	for _, row := range rows {
		for i, v := range row {
			delta := v - means[i]
			scales[i] += delta * delta
		}
	}
	for i := range scales {
		scales[i] = math.Sqrt(scales[i] / float64(len(rows)))
		if scales[i] == 0 {
			scales[i] = 1.0
		}
	}

	standardized = make([][]float64, 0, len(rows))
	for _, row := range rows {
		out := make([]float64, len(row))
		for i, v := range row {
			out[i] = (v - means[i]) / scales[i]
		}
		standardized = append(standardized, out)
	}
	return means, scales, standardized
}

func fitLogisticRegression(rows [][]float64, labels []float64, learningRate float64, epochs int, l2 float64) (weights []float64, bias float64) {
	featureCount := 0
	if len(rows) > 0 {
		featureCount = len(rows[0])
	}
	weights = make([]float64, featureCount)
	if len(rows) == 0 {
		return weights, 0
	}

	for epoch := 0; epoch < epochs; epoch++ {
		gradient := make([]float64, featureCount)
		biasGradient := 0.0

		for i, row := range rows {
			prediction := sigmoid(dot(weights, row) + bias)
			err := prediction - labels[i]
			for j, v := range row {
				gradient[j] += err * v
			}
			biasGradient += err
		}

		rowCount := float64(len(rows))
		for j := range weights {
			regularised := gradient[j]/rowCount + l2*weights[j]
			weights[j] -= learningRate * regularised
		}
		bias -= learningRate * (biasGradient / rowCount)
	}
	return weights, bias
}

func evaluateMetrics(dataset *trainingDataset, weights []float64, bias float64) []models.MetricValue {
	var truePositive, trueNegative, falsePositive, falseNegative, logLoss float64
	for i, row := range dataset.Rows {
		probability := sigmoid(dot(weights, row) + bias)
		if probability < 1e-6 {
			probability = 1e-6
		} else if probability > 1.0-1e-6 {
			probability = 1.0 - 1e-6
		}
		predicted := 0.0
		if probability >= 0.5 {
			predicted = 1.0
		}
		label := dataset.Labels[i]
		logLoss += -(label*math.Log(probability) + (1.0-label)*math.Log(1.0-probability))
		switch {
		case predicted == 1.0 && label == 1.0:
			truePositive++
		case predicted == 0.0 && label == 0.0:
			trueNegative++
		case predicted == 1.0 && label == 0.0:
			falsePositive++
		case predicted == 0.0 && label == 1.0:
			falseNegative++
		}
	}

	total := truePositive + trueNegative + falsePositive + falseNegative
	if total < 1.0 {
		total = 1.0
	}
	accuracy := roundMetric((truePositive + trueNegative) / total)

	pdenom := truePositive + falsePositive
	if pdenom < 1.0 {
		pdenom = 1.0
	}
	precision := roundMetric(truePositive / pdenom)

	rdenom := truePositive + falseNegative
	if rdenom < 1.0 {
		rdenom = 1.0
	}
	recall := roundMetric(truePositive / rdenom)

	var f1 float64
	if precision+recall > 0 {
		f1 = roundMetric(2.0 * precision * recall / (precision + recall))
	}

	return []models.MetricValue{
		{Name: "accuracy", Value: accuracy},
		{Name: "precision", Value: precision},
		{Name: "recall", Value: recall},
		{Name: "f1", Value: f1},
		{Name: "log_loss", Value: roundMetric(logLoss / total)},
	}
}

func selectMetric(metrics []models.MetricValue, name string) *models.MetricValue {
	for i := range metrics {
		if metrics[i].Name == name {
			m := metrics[i]
			return &m
		}
	}
	return nil
}

func dot(left, right []float64) float64 {
	var sum float64
	n := len(left)
	if len(right) < n {
		n = len(right)
	}
	for i := 0; i < n; i++ {
		sum += left[i] * right[i]
	}
	return sum
}

func sigmoid(v float64) float64 {
	return 1.0 / (1.0 + math.Exp(-v))
}

func roundMetric(v float64) float64 {
	return math.Round(v*10_000.0) / 10_000.0
}
